fix: apply PBAC to routing form CRUD

eunjae-lee opened the pull request:
## What does this PR do?

This PR applies PBAC to routing form CRUD.


## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

eunjae-lee committed:
fix: apply PBAC to routing form CRUD

eunjae-lee committed:
apply permission checks to the UI

eunjae-lee committed:
moving prisma call to repository [WIP]

eunjae-lee committed:
rename repository and fix type error

eunjae-lee committed:
update implementation

eunjae-lee committed:
Merge branch 'main' into fix/pbac-routing-form

eunjae-lee committed:
Merge branch 'main' into fix/pbac-routing-form

eunjae-lee committed:
fix formMutation handler

eunjae-lee committed:
remove unused import

eunjae-lee committed:
Merge branch 'main' into fix/pbac-routing-form

eunjae-lee committed:
Merge branch 'main' into fix/pbac-routing-form

eunjae-lee committed:
revert some rename

hariombalhara commented on packages/lib/server/repository/RoutingFormRepository.ts:
We should call it PrismaRoutingFormRespository

hariombalhara commented on packages/lib/server/repository/RoutingFormRepository.ts:
```diff
@@ -1,6 +1,36 @@
+    return (await prisma.app_RoutingForms_Form.findUnique({
+      where: { id },
+      select,
+    })) as Promise<Prisma.App_RoutingForms_FormGetPayload<{ select: T }> | null>;
```
Probably a follow-up but we should return a non-prisma type and also extend this repository from RoutinFormRespository.interface

hariombalhara reviewed (dismissed):
Overall looks good !! I am yet to test but unit tests are failing right now

hariombalhara commented on packages/app-store/routing-forms/components/getServerSidePropsSingleForm.ts:
There is still usage of isFormCreateEditAllowed in this file https://github.com/calcom/cal.com/blob/fix/pbac-routing-form/packages/app-store/routing-forms/components/getServerSidePropsSingleForm.ts#L37


I would like to do away with that fn now

eunjae-lee committed:
add RolePermission for 'routingForm'

eunjae-lee committed:
Revert "revert some rename"

This reverts commit 0ef3114c0d27b821271b36296de531654b37749f.

eunjae-lee committed:
clean up PrismaRoutingFormRepository

eunjae-lee committed:
fix unit test

eunjae-lee committed:
remove no longer necessary code

eunjae-lee commented on packages/app-store/routing-forms/components/getServerSidePropsSingleForm.ts:
```diff
@@ -30,16 +34,6 @@ export const getServerSidePropsForSingleFormView = async function getServerSideP
-      },
-    };
-  }
-
   const form = await prisma.app_RoutingForms_Form.findUnique({
```
I've removed this code block. I think it's no longer necessary because below these lines, we return if this user is able to create or edit Routing form. And the flags will be used in the UI and the actual `formMutation.handler.ts` will finally check if user can do it or not. So I think this create or edit permission check is not really needed here.

What do you think? @hariombalhara

eunjae-lee committed:
fix type definition

hariombalhara commented on packages/app-store/routing-forms/pages/reporting/[...appPages].tsx:
```diff
@@ -266,6 +266,7 @@ const Reporter = ({ form }: { form: inferSSRProps<typeof getServerSideProps>["fo
 };
 
 export default function ReporterWrapper({
```
I think we don't use this reporting page anymore. It is superseded by Insights. We should remove this.


'View Responses' CTA in Routing Form takes to Insights page and not this reporting page.

hariombalhara replied:
Opened a followup ticket

hariombalhara commented on packages/app-store/routing-forms/components/getServerSidePropsSingleForm.ts:
```diff
@@ -30,16 +34,6 @@ export const getServerSidePropsForSingleFormView = async function getServerSideP
     };
   }
 
-  const isFormCreateEditAllowed = (await import("../lib/isFormCreateEditAllowed")).isFormCreateEditAllowed;
-  if (!(await isFormCreateEditAllowed({ userId: user.id, formId, targetTeamId: null }))) {
-    return {
-      redirect: {
-        permanent: false,
-        destination: "/403",
-      },
-    };
-  }
-
```
Now that we don't show 404 for a Team Routing Form to a regular member, we should allow the user to navigate to this Routing Form

Right now user can't navigate to it. 
<img width="2964" height="448" alt="image" src="https://github.com/user-attachments/assets/f9d7125b-618a-4f48-baed-90239c61163e" />

With the changes in the PR, save and Delete buttons are correctly disabled
<img width="3300" height="766" alt="image" src="https://github.com/user-attachments/assets/2c551f71-ebb6-4a8d-899a-3065d19be346" />

hariombalhara replied:
Opened a followup ticket

hariombalhara committed:
Merge branch 'main' into fix/pbac-routing-form

sean-brydon commented on packages/prisma/migrations/20250806093054_insert_role_permissions_for_routing_form_pbac/migration.sql:
```diff
@@ -0,0 +1,23 @@
+-- Add routing form permissions for admin role (create, read, update, delete)
+INSERT INTO "RolePermission" (id, "roleId", resource, action, "createdAt")
+SELECT
+  gen_random_uuid(), 'admin_role', resource, action, NOW()
+FROM (
+  VALUES
+    ('routingForm', 'create'),
+    ('routingForm', 'read'),
+    ('routingForm', 'update'),
+    ('routingForm', 'delete'),
+    ('routingForm', 'manage')
+) AS permissions(resource, action)
+ON CONFLICT ("roleId", resource, action) DO NOTHING;
+
+-- Add routing form read permission for member role
+INSERT INTO "RolePermission" (id, "roleId", resource, action, "createdAt")
+SELECT
+  gen_random_uuid(), 'member_role', resource, action, NOW()
+FROM (
+  VALUES
+    ('routingForm', 'read')
+) AS permissions(resource, action)
+ON CONFLICT ("roleId", resource, action) DO NOTHING;
```
@hariombalhara Im pretty sure thats not needed here. PBAC is all behind a feature flag and this just seeds the default roles ( which arent actually used unless the feature flag is enabled)

eunjae-lee committed:
explicit permission handling

eunjae-lee committed:
do not disable un-editable routing form

eunjae-lee commented on packages/app-store/routing-forms/components/getServerSidePropsSingleForm.ts:
```diff
@@ -105,12 +105,53 @@ export const getServerSidePropsForSingleFormView = async function getServerSideP
+    canRead: true,
+    canEdit: true,
+    canDelete: true,
+  };
```
applied your feedback!

devin-ai-integration[bot] committed:
fix: correct property name from readonly to readOnly in ListLinkItem

- Fix TypeScript error in routing forms component
- Change readonly={readOnly} to readOnly={readOnly} to match component interface
- Resolves type check failure in CI

Co-Authored-By: hariom@cal.com <hariombalhara@gmail.com>

eunjae-lee committed:
Merge branch 'main' into fix/pbac-routing-form

emrysal reviewed (approved):
Looks good 👍
